### PR TITLE
rsx: Restructure present code

### PR DESCRIPTION
**Pull request by kd-11**
Reorganizes present code for both backends to support some expansion I have been working on and will be added soon. Also fixes a long-standing regression with some games (e.g newer R&C series games) where WCB is needed for output to show on screen while eating up a chunk of performance. (https://github.com/RPCS3/rpcs3/issues/4258)

### CODE REVIEW

**Comment by aka2k** [comment]
Gran Turismo 5 & 6 has no image with this PR. Everything else seems to be working.

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/4079581/RPCS3.log.gz)

**Comment by SpeddieSwagetti** [comment]
This pull also fixed poor upscaling behavior in R&C:ToD. Good stuff!

**Comment by Xcedf** [comment]
Confirm now GT5 has smth R&C had
WCB Off balckscreen output
WCB On

**Comment by kd-11** [comment]
Should be fixed now

File: rpcs3/Emu/RSX/GL/GLPresent.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+	rsx::thread::flip(info);
+
+	// Cleanup
+	m_gl_texture_cache.on_frame_end();
+	m_vertex_cache->purge();
+
+	auto removed_textures = m_rtts.free_invalidated();
+	m_framebuffer_cache.remove_if([&](auto& fbo)
+	{
+		if (fbo.unused_check_count() >= 2) return true; // Remove if stale
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLPresent.cpp]: Could do this in a single return

File: rpcs3/Emu/RSX/GL/GLPresent.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+				info->width = rsx::apply_resolution_scale(std::min(surface_width, static_cast<u16>(info->width)), true);
+				info->height = rsx::apply_resolution_scale(std::min(surface_height, static_cast<u16>(info->height)), true);
+			}
+		}
+	}
+	else if (auto surface = m_gl_texture_cache.find_texture_from_dimensions<true>(info->address, info->format, info->width, info->height))
+	{
+		//Hack - this should be the first location to check for output
+		//The render might have been done offscreen or in software and a blit used to display
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLPresent.cpp]: Format

**Comment by Xcedf** [comment]
Yea, fixed.

File: rpcs3/Emu/RSX/VK/VKPresent.cpp [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+		vk::change_image_layout(*m_current_command_buffer, target_image, target_layout, present_layout, subresource_range);
+	}
+
+	queue_swap_request();
+
+	m_frame_stats.flip_time = m_profiler.duration();
+
+	m_frame->flip(m_context);
+	rsx::thread::flip(info);
+}
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKPresent.cpp]: Newline

**Comment by Asinin3** [comment]
Just wanted to mention, Jenci said on Discord that NFS The Run doesn't need Write Color Buffers anymore either. If anyone has other games that need WCB, they should check if they still do. And mention your findings here.

**Comment by legend800** [comment]
**Everything that needed WCB before still needs it in my collection:**
MS1: black screens
MS Rift: bloom
MS Apoc: bloom
Peggle: flickering gfx
WipeoutHD: bloom
Time Crisis Razing Storm: bloom
Deadstorm Pirates: physics

**Untested:**
Gow3
GowA
WRC2
WRC3
Split Second

**Comment by Asinin3** [comment]
It might just be a frostbite thing.

**Comment by SpeddieSwagetti** [comment]
Nier: Replicant still needs WCB to fix the white screen ingame.

**Comment by kd-11** [comment]
WCB is not a bug with rare exceptions. Most games that needed it before will forever need it. PS3 provides this functionality just fine and games are free to use it as they see fit.

**Pull request merged**